### DISCUSSION TITLE
Issue with anim_to_vtk and ParaView 12?

### DISCUSSION THREAD

**Discussion by aperijake** [OpenRadioss/OpenRadioss#2787] [Q&A]
I am wondering if anyone has noticed an issue with viewing vtk files produced by anime_to_vtk in ParaView 12 or later. I simulated a cylinder with tet elements, and the results look like elements are missing (see image). After some troubleshooting, I found that ParaView 11.2 looks good, but 12.0, 12.1, and 13 RC1 have the missing elements.
<img width="889" alt="missing_tets" src="https://github.com/OpenRadioss/OpenRadioss/assets/146987853/d2a482d0-929c-4a85-b22e-1f27e08fd1d2">

**Comment by Vortex-CAE** [comment]
@MarianBulla we had a similar issue in the early days of vortex-radioss. It is probably due to the order in which the node connectivity is applied on degenerate solids.

**Reply by aperijake** [reply]
@MarianBulla  Thanks for the insight! I am new to OpenRadioss, so I have a lot to learn. I don't use tet4s much due to the locking issues, but it is good to know the caveats when I work with them.  Am I correct in assuming tet4s are treated as degenerate hex8s throughout the code?

I wrote a quick Python script to convert the degenerate hexes to tets in the vtk file (below). That fixed the issue for the cylinder example. I glanced at `tools/anim_to_vtk/src/anim_to_vtk.cpp` to see if the logic could be added there, but I think this will take more time than I can afford to allocate to it now. Would this be a worthy enhancement request? Tagging @sebastienVilleneuve as it looks like you have worked a lot on `anim_to_vtk`.

```python
import argparse

def convert_degenerate_hexes_to_tets(input_file, output_file):
    with open(input_file, 'r') as f:
        lines = f.readlines()

    # Open the output file for writing
    with open(output_file, 'w') as f:
        line_iterator = iter(lines)

        for line in line_iterator:
            if line.startswith('CELLS'):
                break
            f.write(line)

        # Write the new CELLS line with correct number of tetrahedra and points
        cells_info = line.split()
        num_hexes = int(cells_info[1])
        num_entries = num_hexes * 5
        f.write(f'CELLS {num_hexes} {num_entries}\n')

        # Convert each hex to 1 tet by removing redundant nodes
        for _ in range(num_hexes):
            line = next(line_iterator)
            if line.startswith('8'):
                nodes = line.split()[1:]
                # Only keep the first four nodes, remove redundant ones
                n0, n2, n4, n6 = nodes[0], nodes[2], nodes[4], nodes[6]
                f.write(f'4 {n0} {n2} {n4} {n6}\n')

        # Write the CELL_TYPES line with correct number of tetrahedral cell types
        for line in line_iterator:
            if line.startswith('CELL_TYPES'):
                f.write(f'CELL_TYPES {num_hexes}\n')
                break
            f.write(line)

        for _ in range(num_hexes):
            f.write('10\n')  # 10 is the VTK cell type for tetrahedra
            line = next(line_iterator)  # skip the line with '12'

        # Continue writing the rest of the file
        for line in line_iterator:
            f.write(line)

def main():
    parser = argparse.ArgumentParser(description="Convert degenerate hex elements to tetrahedral elements in a VTK file.")
    parser.add_argument("input_file", help="Path to the input VTK file")
    parser.add_argument("output_file", help="Path to the output VTK file")
    
    args = parser.parse_args()
    convert_degenerate_hexes_to_tets(args.input_file, args.output_file)

if __name__ == "__main__":
    main()
```

**Reply by MarianBulla** [reply]
Hi @aperijake,

thank you very much for this !

We also discussed this issue with one person from kitware. 

It comes exactly from the point how the tetra elements are described internally, using the "degenerated" hexa elements.
It should be fixed, in the latest anim2vtk converter.

Thank you very much again for this hint !
Much appreciated.

Thanks and best regards,
Marian

**Reply by aperijake** [reply]
I'm glad the hint helped @MarianBulla! Thanks for taking a look at the issue.
Jake

**Comment by sebastienVilleneuve** [comment]
@MarianBulla  @aperijake 

Yes it will be fixed soon , but not yet merged ( PR is created)
By the way , you can recompile the PR branch easily to test the fix

Hope this helps
Sébastien

**Reply by aperijake** [reply]
Thanks for getting the fix in @sebastienVilleneuve. I will verify that it works on my problem soon.
